README.md
# Beaker Host Generator

`beaker-hostgenerator` is a command line utility designed to generate beaker
host config files using a compact command line SUT specification.

<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-generate-toc again -->
**Table of Contents**

- [Beaker Host Generator](#beaker-host-generator)
    - [Hypervisors](#hypervisors)
    - [Usage](#usage)
        - [Simple two-host layout](#simple-two-host-layout)
        - [Single host with Arbitrary Roles](#single-host-with-arbitrary-roles)
        - [Two hosts with multiple hypervisors and arbitrary host settings](#two-hosts-with-multiple-hypervisors-and-arbitrary-host-settings)
        - [Two hosts with arbitrary host settings with arbitrary lists](#two-hosts-with-arbitrary-host-settings-with-arbitrary-lists)
        - [Arbitrary nested hashes and arrays](#arbitrary-nested-hashes-and-arrays)
        - [Arbitrary global configuration settings](#arbitrary-global-configuration-settings)
        - [Custom hypervisor](#custom-hypervisor)
        - [URL-encoded input](#url-encoded-input)
    - [Testing](#testing)
        - [Test Fixtures](#test-fixtures)
            - [Generated Fixtures](#generated-fixtures)
    - [Support](#support)
    - [License](#license)

<!-- markdown-toc end -->

## Hypervisors

Any hypervisor may be specified and generated, but if it's not a built-in
hypervisor you will have to provide the entire hypervisor configuration as
input. See the [Custom hypervisor](#custom-hypervisor) example for more
information.

It currently provides built-in configuration for Puppets' internal
[vmpooler][vmpooler] hypervisor, [always-be-scheduling][always-be-scheduling]
hypervisor, static (non-provisioned) nodes, and is designed in a way that makes
it possible to easily add support for additional hypervisors
(any hypervisor type supported by [beaker][beaker]).

To see the list of built-in hypervisors you can run:
```
$ beaker-hostgenerator --list
```

## Usage

Below are some example usages of `beaker-hostgenerator`.

### Simple two-host layout

```
$ beaker-hostgenerator centos6-64mdca-32a
```

Will generate

```yaml
---
HOSTS:
  centos6-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-6-x86_64
    template: centos-6-x86_64
    roles:
    - agent
    - master
    - database
    - dashboard
  centos6-32-2:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-6-i386
    template: centos-6-i386
    roles:
    - agent
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
```

### Single host with Arbitrary Roles

```
$ beaker-hostgenerator centos6-32compile_master,another_role.ma
```

Will generate

```yaml
---
HOSTS:
  centos6-32-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-6-i386
    template: centos-6-i386
    roles:
    - agent
    - master
    - compile_master
    - another_role
    frictionless_options:
      main:
        dns_alt_names: puppet
        environmentpath: "/etc/puppetlabs/puppet/environments"
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
```

### Two hosts with multiple hypervisors and arbitrary host settings

```
$ beaker-hostgenerator centos6-64m{hypervisor=none\,hostname=static-master}-redhat7-64a{somekey=some-value}
```

Will generate

```yaml
---
HOSTS:
  static-master:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    platform: el-6-x86_64
    hypervisor: none
    roles:
    - agent
    - master
  redhat7-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-7-x86_64
    template: redhat-7-x86_64
    somekey: some-value
    roles:
    - agent
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
```

### Two hosts with arbitrary host settings with arbitrary lists

```
$ beaker-hostgenerator centos6-64m{disks=\[16\]}-redhat7-64a{disks=\[8\,16\]}
```

Will generate

```yaml
---
HOSTS:
  centos6-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    platform: el-6-x86_64
    hypervisor: vmpooler
    disks:
    - 16
    roles:
    - agent
    - master
  redhat7-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-7-x86_64
    template: redhat-7-x86_64
    disks:
    - 8
    - 16
    roles:
    - agent
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
```

### Arbitrary nested hashes and arrays

```
$ beaker-hostgenerator --global-config {host_tags={lifetime=4h}\,list=[{listkey=listvalue}]\} redhat7-64m{hostlist=\[string\,{key=value}\,1234\]}
```

Will generate

```yaml
---
HOSTS:
  redhat7-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-7-x86_64
    template: redhat-7-x86_64
    hostlist:
    - string
    - key: value
    - 1234
    roles:
    - agent
    - master
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
  host_tags:
    lifetime: 4h
  list:
  - listkey: listvalue
```

### Arbitrary global configuration settings

```
$ beaker-hostgenerator --global-config {preserve_hosts=onfail\,log_level=debug\,server.ip=12.345.6789} redhat7-64m
```

Will generate

```yaml
---
HOSTS:
  redhat7-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-7-x86_64
    template: redhat-7-x86_64
    roles:
    - agent
    - master
CONFIG:
  nfs_server: none
  consoleport: 443
  preserve_hosts: onfail
  log_level: debug
  server.ip: 12.345.6789
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
```

### Custom hypervisor

The following example shows one way of generating a custom hypervisor that
includes both per-host configuration and global configuration.

The term "custom" in this case signifies that it's not a built-in hypervisor
(like `vmpooler`), which means we'll have to provide all the configuration
ourselves as there isn't any built-in configuration for our hypervisor.

```
$ beaker-hostgenerator --hypervisor=custom --global={custom_api=http://api.custom.net} centos6-64
```

Will generate

```yaml
---
HOSTS:
  centos6-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    platform: el-6-x86_64
    hypervisor: custom
    roles:
    - agent
CONFIG:
  nfs_server: none
  consoleport: 443
  custom_api: http://api.custom.net
```

### URL-encoded input

It may be necessary to URL-encode the input in order for it to properly be used
in certain contexts, such as Jenkins.

In most cases it will only be necessary to escape the characters that support
arbitrary settings, which means the following characters:

- `{` is `%7B`
- `,` is `%2C`
- `}` is `%7D`
- ` ` is `%20`
- `[` is `%5B`
- `]` is `%5D`

For a full URL encoding reference see: http://www.w3schools.com/tags/ref_urlencode.asp

```
$ beaker-hostgenerator centos6-64mcd-aix53-POWERfa%7Bhypervisor=aix%2Cvmhostname=pe-aix-53-acceptance.delivery.puppetlabs.net%7D
```

Is equivalent to

```
$ beaker-hostgenerator centos6-64mcd-aix53-POWERfa{hypervisor=aix,vmhostname=pe-aix-53-acceptance.delivery.puppetlabs.net}
```

And will generate

```yaml
---
HOSTS:
  centos6-64-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    hypervisor: vmpooler
    platform: el-6-x86_64
    template: centos-6-x86_64
    roles:
    - agent
    - master
    - dashboard
    - database
  aix53-POWER-1:
    pe_dir:
    pe_ver:
    pe_upgrade_dir:
    pe_upgrade_ver:
    platform: aix-5.3-power
    hypervisor: aix
    vmhostname: pe-aix-53-acceptance.delivery.puppetlabs.net
    roles:
    - agent
    - frictionless
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
```

## Testing

Beaker Host Generator currently uses both rspec and minitest tests. To run both
at the same time, run:
```bash
bundle exec rake test
```

### Test Fixtures

Beaker Host Generator makes extensive use of test fixtures to validate its
behavior under specific conditions. An example of such a test fixture is as
follows:

```yaml
---
arguments_string: "--pe_dir /opt/hello centos6-64mdc"
environment_variables: {}
expected_hash:
  HOSTS:
    centos6-64-1:
      pe_dir: "/opt/hello"
      pe_ver: 
      pe_upgrade_dir: 
      pe_upgrade_ver: 
      hypervisor: vmpooler
      platform: centos-6-x86_64
      template: centos-6-x86_64
      roles:
      - agent
      - master
      - database
      - dashboard
  CONFIG:
    nfs_server: none
    consoleport: 443
    pooling_api: http://vmpooler.delivery.puppetlabs.net/
expected_exception: 
```

These test fixtures are yaml files searched for in the directory
`test/fixtures`. The data structure expected in these files is a hash with four
keys:

- `arguments_string`: The command line arguments that should be passed to
  `beaker-hostgenerator`
- `environment_variables`: The environment variables that should be set during
  the `beaker-hostgenerator` call.
- `expected_hash`: A hash that should match the output of `beaker-hostgenerator`
  when it is run with `options\_string`
- `expected_exception`: If the `arguments_string` passed to `beaker-hostgenerator`
  is expected to lead to an exceptional state, this is the name of the exception
  that the fixture test will attempt to match.

#### Generated Fixtures

It is possible to generate test fixtures using the current state of the
`beaker-hostgenerator` library. To do this, call the `generate:fixtures` Rake
task.

However, this is not something that should need to be done very often. If you
are running tests and find that some fixtures no longer work, you have most
likely made a change that incompatibly changes the behavior of
`beaker-hostgenerator` for other users. Use the test fixtures as a guide to
figure out what you did wrong and figure out how to achieve your goal without
potentially breaking `beaker-hostgenerator` for other users.

There are a few circumstances when you should expect to run the
`generate:fixtures` task:

- When you modify the `FixtureGenerator` to generate new fixtures.
- When you need to fix a bug (generated hosts are not usable without your
  change, for example).
- When preparing for a major version bump of Beaker Host Generator.


## Support

Support offered by [Puppet](https://puppet.com) may not always be timely
since it is maintained by a tooling support team that is primarily focused on
improving tools, infrastructure, and automation for our Enterprise products.

That being said, we will happily accept and review PRs from community members
interested in extending and using `beaker-hostgenerator` for their own purposes.
See the [contributing][contributing] doc for more information about how to
contribute.

If you have questions or comments, please contact the Beaker team at the
`#puppet-dev` IRC channel on chat.freenode.org

## Make a new release

To make a new release:

* Switch to a new git branch
* Update `lib/beaker-hostgenerator/version.rb` to the new version
* export `CHANGELOG_GITHUB_TOKEN` with a GitHub access token
* run `bundle exec rake changelog`
* Review the CHANGELOG.md
* Create a PR
* Get it reviewed and merged
* Create a signed git tag with the same version number and push it
* GitHub actions will do the release

## License

`beaker-hostgenerator` is distributed under the
[Apache License, Version 2.0][apache-v2]. See the [LICENSE][license] file for more details.

[vmpooler]: https://github.com/puppetlabs/vmpooler
[beaker]: https://github.com/puppetlabs/beaker
[license]: LICENSE
[contributing]: CONTRIBUTING.md
[apache-v2]: http://www.apache.org/licenses/LICENSE-2.0.html
[always-be-scheduling]: https://github.com/puppetlabs/always-be-scheduling


HISTORY.md
## [1.1.26](https://github.com/voxpupuli/modulesync/tree/1.1.26) (2019-02-19)
- Add platform
  - debian10-64
  - debian10-32
- Change amazon7-ARM64 to use el-7-aarch64 platform

## [1.1.25] - 2019-01-07
- Fix --hypervisor vagrant_libvirt

## [1.1.24] - 2019-01-03
- Add platform:
  - amazon7-ARM64

## [1.1.23] - 2018-12-13
- Add platforms:
  - redhat8-64
  - osx1014-64

## [1.1.22] - 2018-11-21
- Add vagrant hypervisor for debian plaforms

## [1.1.21] - 2018-11-19
- Disable getty when using docker

## [1.1.20] - 2018-11-14
- Add gnupg to docker_image_commands for debian9-64
- Add platform:
  - fedora29-64

## [1.1.19] - 2018-10-18
- Add platforms:
  - solaris114-32
  - solaris114-64

## [1.1.18] - 2018-10-10
- Add platform:
  - ubuntu1810-64

## [1.1.17] - 2018-10-05
- Add platforms:
  - windows2019_core-64
  - windows2019_core-6432

## [1.1.16] - 2018-09-04
- Install iproute2 on ubuntu18.04 for containers

## [1.1.15] - 2018-08-16
- Add platforms:
  - panos61-64
  - panos71-64
## Details

### 0.9.0 - 7 Jun, 2017 (06b6a0da)

* (GEM) update beaker-hostgenerator version to 0.9.0 (06b6a0da)

* Merge pull request #80 from puppetlabs/prepare-for-release (b7a83889)


```
Merge pull request #80 from puppetlabs/prepare-for-release

(MAINT) Update CHANGELOG for 0.9.0 release
```
* (MAINT) Update CHANGELOG for 0.9.0 release (aae2e698)

* (MAINT) Fix rendering issue in README and update TOC (bc92bd07)

* (QENG-4945) Add support for arbitrary lists (e75f1e7e)


```
(QENG-4945) Add support for arbitrary lists

Before, we had no way to add lists to configurations. This adds
a new bracket syntax within arbitrary settings in order to parse out any lists the user may want to
add in.
```
* Merge pull request #79 from puppetlabs/qeng-5041 (990b6081)


```
Merge pull request #79 from puppetlabs/qeng-5041

(QENG-5041) Add vRO to support platforms
```
* (QENG-5041) Add vRO to support platforms (22b7c670)


```
(QENG-5041) Add vRO to support platforms

This adds both vro-6-x86_64 and vro-7-x86_64. The templates
already exist within vmpooler
```
* Merge pull request #77 from puppetlabs/fix-default-config-hypervisors (988476e1)


```
Merge pull request #77 from puppetlabs/fix-default-config-hypervisors

(maint) The hypervisor defaults are not overwritable
```
* Merge pull request #76 from puppetlabs/qeng4965/add-abs-hardware-support (874f473f)


```
Merge pull request #76 from puppetlabs/qeng4965/add-abs-hardware-support

(QENG-4965) Add ABS support for hardware platforms
```
* (MAINT) Delete duplicate keys from test helper hash (0fe2c0ac)

* (MAINT) Prefer underscores over dashes in file names (efe8b7b1)


```
(MAINT) Prefer underscores over dashes in file names

Ruby prefers that you not use dashes in your .rb filenames, so this commit
renames abs-support.rb to abs_support.rb
```
* (maint) The hypervisor defaults are not overwritable (1a44141b)


```
(maint) The hypervisor defaults are not overwritable
Before this change the code would pull and merge the hypervisor defaults on top of any
user provided --global-config block. This reverses the logic to use defaults only
if the key is not specified in the --global-config section. This specifically
enables overwritting the pooling_api for the vmpooler hypervisor, as it is the
only one with a default global_config() method implementation. Also added it's
respective test
```
* (QENG-4965) Add ABS support for hardware platforms (a9ddf44b)


```
(QENG-4965) Add ABS support for hardware platforms

This commit makes the hardware platforms compatible with the ABS hypervisor and
the --templates-only switch to enable hardware support in Jenkins via ABS.
```
### <a name = "0.8.4">0.8.4 - 6 Apr, 2017 (77943fb2)

* (HISTORY) update beaker-hostgenerator history for gem release 0.8.4 (77943fb2)

* (GEM) update beaker-hostgenerator version to 0.8.4 (d80eb358)

* Merge pull request #74 from puppetlabs/prepare-for-release (43405262)


```
Merge pull request #74 from puppetlabs/prepare-for-release

(MAINT) Update CHANGELOG for 0.8.4 release
```
* (MAINT) Update CHANGELOG for 0.8.4 release (26cd2eef)

* Merge pull request #73 from glennsarti/2012r2-wmf5 (d00a1d63)


```
Merge pull request #73 from glennsarti/2012r2-wmf5

(MODULES-4667) Add Windows 2012R2 WMF5 VMPooler image
```
* (MODULES-4667) Add Windows 2012R2 WMF5 VMPooler image (8b293890)


```
(MODULES-4667) Add Windows 2012R2 WMF5 VMPooler image

This commit adds the win-2012r2-wmf5-x86_64 vmpooler image, which is a variation
on the Windows Sever 2012R2 64bit image.
```
### <a name = "0.8.3">0.8.3 - 7 Feb, 2017 (79e8d717)

* (HISTORY) update beaker-hostgenerator history for gem release 0.8.3 (79e8d717)

* (GEM) update beaker-hostgenerator version to 0.8.3 (12bfa089)

* Merge pull request #72 from puppetlabs/maint/prepare-for-release (cc5a3ce6)


```
Merge pull request #72 from puppetlabs/maint/prepare-for-release

(MAINT) Update CHANGELOG for 0.8.3 release
```
* (MAINT) Update CHANGELOG for 0.8.3 release (6d3900a9)

* Merge pull request #71 from ferglor/BKR-1032 (cca29c01)


```
Merge pull request #71 from ferglor/BKR-1032

(BKR-1032) Switched from using STDERR to $stderr
```
* (BKR-1032) Switched from using STDERR to $stderr (7d4f8197)

* Merge pull request #70 from puppetlabs/qeng4019/remove-warning-message (087df5a5)


```
Merge pull request #70 from puppetlabs/qeng4019/remove-warning-message

(QENG-4019) Remove warning message about changing defaults
```
* (QENG-4019) Remove warning message about changing defaults (d3e19a40)


```
(QENG-4019) Remove warning message about changing defaults

This warning message was added a while ago to inform users of an upcoming
breaking change. We're no longer certain that we'll be making the breaking
change though, so the warning is a bit misleading.

Furthermore, we're now trying to silence this message when beaker-hostgenerator
is used programatically within Beaker.

There will likely be more changes that are outlined in the linked ticket, such

lib/genconfig/data/vsphere.rb
module GenConfig
  module Data
    module VSphere

      OSINFO = {
        'centos4-32' => {
          'platform' => 'el-4-i386',
          'template' => 'centos-4-i386'
        },
        'centos4-64' => {
          'platform' => 'el-4-x86_64',
          'template' => 'centos-4-x86_64'
        },
        'centos5-32' => {
          'platform' => 'el-5-i386',
          'template' => 'centos-5-i386'
        },
        'centos5-64' => {
          'platform' => 'el-5-x86_64',
          'template' => 'centos-5-x86_64'
        },
        'centos6-32' => {
          'platform' => 'el-6-i386',
          'template' => 'centos-6-i386'
        },
        'centos6-64' => {
          'platform' => 'el-6-x86_64',
          'template' => 'centos-6-x86_64'
        },
        'centos7-64' => {
          'platform' => 'el-7-x86_64',
          'template' => 'centos-7-x86_64'
        },
        'debian6-32' => {
          'platform' => 'debian-6-i386',
          'template' => 'debian-6-i386'
        },
        'debian6-64' => {
          'platform' => 'debian-6-amd64',
          'template' => 'debian-6-x86_64'
        },
        'debian7-32' => {
          'platform' => 'debian-7-i386',
          'template' => 'debian-7-i386'
        },
        'debian7-64' => {
          'platform' => 'debian-7-amd64',
          'template' => 'debian-7-x86_64'
        },
        'debian8-32' => {
          'platform' => 'debian-8-i386',
          'template' => 'debian-8-i386'
        },
        'debian8-64' => {
          'platform' => 'debian-8-amd64',
          'template' => 'debian-8-x86_64'
        },
        'oracle5-32' => {
          'platform' => 'el-5-i386',
          'template' => 'oracle-5-i386'
        },
        'oracle5-64' => {
          'platform' => 'el-5-x86_64',
          'template' => 'oracle-5-x86_64'
        },
        'oracle6-32' => {
          'platform' => 'el-6-i386',
          'template' => 'oracle-6-i386'
        },
        'oracle6-64' => {
          'platform' => 'el-6-x86_64',
          'template' => 'oracle-6-x86_64'
        },
        'oracle7-64' => {
          'platform' => 'el-7-x86_64',
          'template' => 'oracle-7-x86_64'
        },
        'osx109-64' => {
          'platform' => 'osx-10.9-x86_64',
          'template' => 'osx-109-x86_64'
        },
        'redhat4-32' => {
          'platform' => 'el-4-i386',
          'template' => 'redhat-4-i386'
        },
        'redhat4-64' => {
          'platform' => 'el-4-x86_64',
          'template' => 'redhat-4-x86_64'
        },
        'redhat5-32' => {
          'platform' => 'el-5-i386',
          'template' => 'redhat-5-i386'
        },
        'redhat5-64' => {
          'platform' => 'el-5-x86_64',
          'template' => 'redhat-5-x86_64'
        },
        'redhat6-32' => {
          'platform' => 'el-6-i386',
          'template' => 'redhat-6-i386'
        },
        'redhat6-64' => {
          'platform' => 'el-6-x86_64',
          'template' => 'redhat-6-x86_64'
        },
        'redhat7-64' => {
          'platform' => 'el-7-x86_64',
          'template' => 'redhat-7-x86_64'
        },
        'fedora14-32' => {
          'platform' => 'fedora-14-i386',
          'template' => 'fedora-14-i386'
        },
        'fedora19-32' => {
          'platform' => 'fedora-19-i386',
          'template' => 'fedora-19-i386'
        },
        'fedora19-64' => {
          'platform' => 'fedora-19-x86_64',
          'template' => 'fedora-19-x86_64'
        },
        'fedora20-32' => {
          'platform' => 'fedora-20-i386',
          'template' => 'fedora-20-i386'
        },
        'fedora20-64' => {
          'platform' => 'fedora-20-x86_64',
          'template' => 'fedora-20-x86_64'
        },
        'fedora21-32' => {
          'platform' => 'fedora-21-i386',
          'template' => 'fedora-21-i386'
        },
        'fedora21-64' => {
          'platform' => 'fedora-21-x86_64',
          'template' => 'fedora-21-x86_64'
        },
        'opensuse11-32' => {
          'platform' => 'opensuse-11-i386',
          'template' => 'opensuse-11-i386'
        },
        'opensuse11-64' => {
          'platform' => 'opensuse-11-x86_64',
          'template' => 'opensuse-11-x86_64'
        },
        'scientific5-32' => {
          'platform' => 'el-5-i386',
          'template' => 'scientific-5-i386'
        },
        'scientific5-64' => {
          'platform' => 'el-5-x86_64',
          'template' => 'scientific-5-x86_64'
        },
        'scientific6-32' => {
          'platform' => 'el-6-i386',
          'template' => 'scientific-6-i386'
        },
        'scientific6-64' => {
          'platform' => 'el-6-x86_64',
          'template' => 'scientific-6-x86_64'
        },
        'scientific7-64' => {
          'platform' => 'el-7-x86_64',
          'template' => 'scientific-7-x86_64'
        },
        'sles10-32' => {
          'platform' => 'sles-10-i386',
          'template' => 'sles-10-i386'
        },
        'sles10-64' => {
          'platform' => 'sles-10-x86_64',
          'template' => 'sles-10-x86_64'
        },
        'sles11-32' => {
          'platform' => 'sles-11-i386',
          'template' => 'sles-11-i386'
        },
        'sles11-64' => {
          'platform' => 'sles-11-x86_64',
          'template' => 'sles-11-x86_64'
        },
        'sles12-64' => {
          'platform' => 'sles-12-x86_64',
          'template' => 'sles-12-x86_64'
        },
        'solaris10-64' => {
          'platform' => 'solaris-10-i386',
          'template' => 'solaris-10-x86_64'
        },
        'solaris11-64' => {
          'platform' => 'solaris-11-i386',
          'template' => 'solaris-11-x86_64'
        },
        'solaris112-64' => {
          'platform' => 'solaris-11.2-i386',
          'template' => 'solaris-112-x86_64'
        },
        'ubuntu1004-32' => {
          'platform' => 'ubuntu-10.04-i386',
          'template' => 'ubuntu-1004-i386'
        },
        'ubuntu1004-64' => {
          'platform' => 'ubuntu-10.04-amd64',
          'template' => 'ubuntu-1004-x86_64'
        },
        'ubuntu1204-32' => {
          'platform' => 'ubuntu-12.04-i386',
          'template' => 'ubuntu-1204-i386'
        },
        'ubuntu1204-64' => {
          'platform' => 'ubuntu-12.04-amd64',
          'template' => 'ubuntu-1204-x86_64'
        },
        'ubuntu1404-32' => {
          'platform' => 'ubuntu-14.04-i386',
          'template' => 'ubuntu-1404-i386'
        },
        'ubuntu1404-64' => {
          'platform' => 'ubuntu-14.04-amd64',
          'template' => 'ubuntu-1404-x86_64'
        },
        'windows2003-64' => {
          'platform' => 'windows-2003-64',
          'template' => 'win-2003-x86_64'
        },
        'windows2003r2-32' => {
          'platform' => 'windows-2003r2-32',
          'template' => 'win-2003r2-i386'
        },
        'windows2003r2-64' => {
          'platform' => 'windows-2003r2-64',
          'template' => 'win-2003r2-x86_64',
        },
        'windows2008-64' => {
          'platform' => 'windows-2008-64',
          'template' => 'win-2008-x86_64'
        },
        'windows2008r2-64' => {
          'platform' => 'windows-2008r2-64',
          'template' => 'win-2008r2-x86_64'
        },
        'windows2012-64' => {
          'platform' => 'windows-2012-64',
          'template' => 'win-2012-x86_64'
        },
        'windows2012r2-64' => {
          'platform' => 'windows-2012r2-64',
          'template' => 'win-2012r2-x86_64'
        },
        'windows7-64' => {
          'platform' => 'windows-7-64',
          'template' => 'win-7-x86_64'
        },
        'windows8-64' => {
          'platform' => 'windows-8-64',
          'template' => 'win-8-x86_64'
        },
        'windows81-64' => {
          'platform' => 'windows-8.1-64',
          'template' => 'win-81-x86_64'
        },
        'windowsvista-64' => {
          'platform' => 'windows-vista-64',
          'template' => 'win-vista-x86_64'
        },
      }

      # Prepended to all Templates. This uniquely identifies them, because
      # their names are fairly generic.
      QA_VCENTER_PATH = 'Delivery/Quality Assurance/Templates/vCloud'

      VSPHERE_CONFIG = {
        'HOSTS' => {},
        'CONFIG' => {
          'datastore' => 'instance0',
          'folder' => 'Delivery/Quality Assurance/Enterprise/Dynamic',
          'resourcepool' => 'delivery/Quality Assurance/Enterprise/Dynamic',
          'pooling_api' => 'http://vcloud.delivery.puppetlabs.net/'
        }
      }

    end
  end
end
